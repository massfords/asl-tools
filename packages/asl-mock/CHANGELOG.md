# @asl-tools/asl-mock

## 0.0.1

### Patch Changes

- 152c8b6: Initial migration of packages to a monorepo
- Updated dependencies [152c8b6]
  - @asl-tools/asl-puml@0.0.1
