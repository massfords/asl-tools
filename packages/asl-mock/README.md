# asl-mock
- provide Typescript types for AWS MockConfigFile
- provide function to visualize an execution
