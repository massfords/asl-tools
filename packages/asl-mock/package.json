{
  "name": "@asl-tools/asl-mock",
  "description": "Types and helpers for AWS MockConfigFile",
  "version": "0.0.1",
  "type": "module",
  "exports": {
    ".": "./dist/index.js"
  },
  "bin": {
    "asl-mock": "./dist/asl-mock.js"
  },
  "scripts": {
    "build": "del dist && tsc",
    "lint": "eslint .",
    "test": "vitest --run",
    "clean": "del .turbo dist coverage"
  },
  "keywords": [
    "asl",
    "amazon states language",
    "mock config",
    "testing",
    "aws",
    "step functions",
    "state machine"
  ],
  "author": "Mark Ford",
  "license": "MIT",
  "devDependencies": {
    "@toolchain/eslint-config": "workspace:*",
    "@toolchain/ts-config": "workspace:*",
    "@toolchain/vitest-config": "workspace:*",
    "@types/eslint": "8.44.2",
    "@types/node": "20.3.1"
  },
  "dependencies": {
    "@aws-sdk/client-sfn": "3.395.0",
    "@asl-tools/asl-puml": "workspace:*",
    "asl-validator": "3.8.0",
    "commander": "11.0.0",
    "jsonpath-plus": "7.2.0",
    "prettier": "3.0.2",
    "testcontainers": "10.2.1",
    "vitest": "0.34.2"
  }
}
