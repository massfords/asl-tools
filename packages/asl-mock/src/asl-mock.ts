#!/usr/bin/env node

import fs from 'fs'
import path from 'path'

import { program } from 'commander'

import { emitTestFile, parseTypeScriptFile } from './lib/generate-test.js'
import { setVerboseLogging } from './logger.js'

function doneValid(): void {
  process.exit(0)
}

function fail(message: string): void {
  console.error(message)
  process.exit(2)
}

program
  .description('Generate test runner from mock config file')
  .requiredOption('-i --input <input>', 'path to mock config ts file')
  .requiredOption('-a --asl <asl>', 'relative path to asl src during test')
  .option('-o --out <out>', 'path for emitted test file')
  .option('--no-esm', 'disables esm style imports')
  .option('--no-flat', 'flat es lint is enabled by default')
  .option('-v --verbose', 'enables verbose logging')
  .parse(process.argv)

function outputFileFromInput(input: string): string {
  if (input.endsWith('.mock.ts')) {
    return input.substring(0, input.length - '.mock.ts'.length) + '.test.ts'
  } else {
    return input.substring(0, input.length - '.ts'.length) + '.test.ts'
  }
}

async function generateTestFile(): Promise<void> {
  try {
    const opts: {
      input: string
      asl: string
      out?: string
      esm: boolean
      flat: boolean
      verbose?: boolean
    } = program.opts()

    setVerboseLogging(opts.verbose ?? false)

    const { found, mockConfigTypeArgs, stateMachines, decl } = parseTypeScriptFile(opts.input)
    if (!found) {
      fail('mock config ts file not found')
      return
    }
    if (!stateMachines) {
      fail('stateMachines not found in mock config')
      return
    }
    if (!mockConfigTypeArgs) {
      fail('mockConfigTypeArgs not found in mock config')
      return
    }
    const outputFile = opts.out ?? outputFileFromInput(opts.input)
    if (fs.existsSync(outputFile)) {
      return
    }
    const output = await emitTestFile({
      testCases: Object.values(stateMachines)[0] as string[],
      aslSourcePath: opts.asl,
      mockConfigSrcFile: `./${path.parse(path.basename(opts.input)).name}`,
      mockConfigTypeArgs,
      mockConfig: decl,
      aslTestRunnerPath: '@asl-tools/asl-mock',
      esm: opts.esm,
      flatEslint: opts.flat,
    })
    fs.writeFileSync(outputFile, output, 'utf-8')
    doneValid()
  } catch (e: unknown) {
    fail(`asl-mock exception: ${(e as { message?: string }).message ?? 'no message'}`)
    // fail("asl-mock exception:" + JSON.stringify(e));
  }
}

void (async () => {
  await generateTestFile()
})()
