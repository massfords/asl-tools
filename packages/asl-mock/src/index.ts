export * from './lib/types.js'
export * from './lib/runner.js'
