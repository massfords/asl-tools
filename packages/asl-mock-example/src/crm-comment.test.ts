// Code generated by asl-mock DO NOT EDIT.
import path from 'path'

import { AslTestRunner } from '@asl-tools/asl-mock'
import { afterAll, afterEach, beforeAll, describe, expect, it } from 'vitest'

import type { CustomErrors, StateMachineNames, StateNames, TestCases } from './crm-comment.mock.js'
import { CrmCommentMock, StartMessages } from './crm-comment.mock.js'

describe('tests for crm-comment.asl.json', () => {
  const TIMEOUT = 30 * 1000
  const outdir = path.join(__dirname, '.asl-puml')
  let _err: unknown = null
  let _aslRunner: AslTestRunner<
    StateMachineNames,
    TestCases,
    StateNames,
    string,
    CustomErrors
  > | null = null

  beforeAll(async () => {
    try {
      const dirname = new URL('.', import.meta.url).pathname
      _aslRunner = await AslTestRunner.createRunner<
        StateMachineNames,
        TestCases,
        StateNames,
        string,
        CustomErrors
      >(CrmCommentMock, {
        'crm-comment': path.join(dirname, 'crm-comment.asl.json'),
      })
    } catch (err: unknown) {
      _err = err
    }
  })

  afterEach(() => {
    _aslRunner?.reset()
  })

  afterAll(async () => {
    await _aslRunner?.stop()
  })

  describe('mock config scenarios', () => {
    const afterCompletion = {
      puml: outdir,
      expectTaskSnapshots: true,
      logHistoryEventsOnFailure: true,
    }

    it(
      'scenario HappyPathTest',
      async () => {
        expect.hasAssertions()
        if (_err) {
          console.error('error starting runner', { err: _err })
        }
        await _aslRunner?.execute(
          {
            name: 'crm-comment',
            startMessage: StartMessages['HappyPathTest'],
            scenario: 'HappyPathTest',
            expect,
          },
          afterCompletion,
        )
      },
      TIMEOUT,
    )

    it(
      'scenario NegativeSentimentTest',
      async () => {
        expect.hasAssertions()
        await _aslRunner?.execute(
          {
            name: 'crm-comment',
            startMessage: StartMessages['NegativeSentimentTest'],
            scenario: 'NegativeSentimentTest',
            expect,
          },
          afterCompletion,
        )
      },
      TIMEOUT,
    )

    it(
      'scenario CustomValidationFailedCatchTest',
      async () => {
        expect.hasAssertions()
        await _aslRunner?.execute(
          {
            name: 'crm-comment',
            startMessage: StartMessages['CustomValidationFailedCatchTest'],
            scenario: 'CustomValidationFailedCatchTest',
            expect,
          },
          afterCompletion,
        )
      },
      TIMEOUT,
    )

    it(
      'scenario ValidationExceptionCatchTest',
      async () => {
        expect.hasAssertions()
        await _aslRunner?.execute(
          {
            name: 'crm-comment',
            startMessage: StartMessages['ValidationExceptionCatchTest'],
            scenario: 'ValidationExceptionCatchTest',
            expect,
          },
          afterCompletion,
        )
      },
      TIMEOUT,
    )

    it(
      'scenario RetryOnServiceExceptionTest',
      async () => {
        expect.hasAssertions()
        await _aslRunner?.execute(
          {
            name: 'crm-comment',
            startMessage: StartMessages['RetryOnServiceExceptionTest'],
            scenario: 'RetryOnServiceExceptionTest',
            expect,
          },
          afterCompletion,
        )
      },
      TIMEOUT,
    )
  })
})
