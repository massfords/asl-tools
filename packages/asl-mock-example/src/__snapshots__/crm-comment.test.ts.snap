// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`tests for crm-comment.asl.json > mock config scenarios > scenario CustomValidationFailedCatchTest > Check Address 1`] = `
{
  "FunctionName": "\${CheckAddressFunctionArn}",
  "Payload": {
    "city": "Columbus",
    "state": "OH",
    "street": "123 Main St",
    "zip": "43219",
  },
}
`;

exports[`tests for crm-comment.asl.json > mock config scenarios > scenario CustomValidationFailedCatchTest > Check Identity 1`] = `
{
  "FunctionName": "\${CheckIdentityFunctionArn}",
  "Payload": {
    "email": "jdoe@example.com",
    "ssn": "123-45-6789",
  },
}
`;

exports[`tests for crm-comment.asl.json > mock config scenarios > scenario CustomValidationFailedCatchTest > CustomValidationFailed 1`] = `
{
  "Entries": [
    {
      "Detail": {
        "Message": "Validation Failed",
      },
      "DetailType": "ValidationFailed",
      "EventBusName": "default",
      "Source": "LocalTestingSource",
    },
  ],
}
`;

exports[`tests for crm-comment.asl.json > mock config scenarios > scenario HappyPathTest > Add to FollowUp 1`] = `
{
  "Item": {
    "PK": {
      "S": "jdoe@example.com",
    },
  },
  "TableName": "\${FollowUpTableName}",
}
`;

exports[`tests for crm-comment.asl.json > mock config scenarios > scenario HappyPathTest > Check Address 1`] = `
{
  "FunctionName": "\${CheckAddressFunctionArn}",
  "Payload": {
    "city": "Columbus",
    "state": "OH",
    "street": "123 Main St",
    "zip": "43219",
  },
}
`;

exports[`tests for crm-comment.asl.json > mock config scenarios > scenario HappyPathTest > Check Identity 1`] = `
{
  "FunctionName": "\${CheckIdentityFunctionArn}",
  "Payload": {
    "email": "jdoe@example.com",
    "ssn": "123-45-6789",
  },
}
`;

exports[`tests for crm-comment.asl.json > mock config scenarios > scenario HappyPathTest > CustomerAddedToFollowup 1`] = `
{
  "Entries": [
    {
      "Detail": {
        "EmailAddress": "jdoe@example.com",
        "Message": "Customer Added for follow up",
      },
      "DetailType": "CustomerAdded",
      "EventBusName": "default",
      "Source": "LocalTestingSource",
    },
  ],
}
`;

exports[`tests for crm-comment.asl.json > mock config scenarios > scenario HappyPathTest > DetectSentiment 1`] = `
{
  "LanguageCode": "en",
  "Text": "I am glad to sign-up for this service. Looking forward to different options.",
}
`;

exports[`tests for crm-comment.asl.json > mock config scenarios > scenario NegativeSentimentTest > Check Address 1`] = `
{
  "FunctionName": "\${CheckAddressFunctionArn}",
  "Payload": {
    "city": "Columbus",
    "state": "OH",
    "street": "123 Main St",
    "zip": "43219",
  },
}
`;

exports[`tests for crm-comment.asl.json > mock config scenarios > scenario NegativeSentimentTest > Check Identity 1`] = `
{
  "FunctionName": "\${CheckIdentityFunctionArn}",
  "Payload": {
    "email": "jdoe@example.com",
    "ssn": "123-45-6789",
  },
}
`;

exports[`tests for crm-comment.asl.json > mock config scenarios > scenario NegativeSentimentTest > DetectSentiment 1`] = `
{
  "LanguageCode": "en",
  "Text": "I am glad to sign-up for this service. Looking forward to different options.",
}
`;

exports[`tests for crm-comment.asl.json > mock config scenarios > scenario NegativeSentimentTest > NegativeSentimentDetected 1`] = `
{
  "Entries": [
    {
      "Detail": {
        "Data": {
          "address": {
            "city": "Columbus",
            "state": "OH",
            "street": "123 Main St",
            "zip": "43219",
          },
          "comments": "I am glad to sign-up for this service. Looking forward to different options.",
          "firstname": "Jane",
          "identity": {
            "email": "jdoe@example.com",
            "ssn": "123-45-6789",
          },
          "lastname": "Doe",
        },
        "Message": "Negative Sentiment Detected",
      },
      "DetailType": "NegativeSentiment",
      "EventBusName": "default",
      "Source": "LocalTestingSource",
    },
  ],
}
`;

exports[`tests for crm-comment.asl.json > mock config scenarios > scenario RetryOnServiceExceptionTest > Add to FollowUp 1`] = `
{
  "Item": {
    "PK": {
      "S": "jdoe@example.com",
    },
  },
  "TableName": "\${FollowUpTableName}",
}
`;

exports[`tests for crm-comment.asl.json > mock config scenarios > scenario RetryOnServiceExceptionTest > Check Address 1`] = `
{
  "FunctionName": "\${CheckAddressFunctionArn}",
  "Payload": {
    "city": "Columbus",
    "state": "OH",
    "street": "123 Main St",
    "zip": "43219",
  },
}
`;

exports[`tests for crm-comment.asl.json > mock config scenarios > scenario RetryOnServiceExceptionTest > Check Identity 1`] = `
{
  "FunctionName": "\${CheckIdentityFunctionArn}",
  "Payload": {
    "email": "jdoe@example.com",
    "ssn": "123-45-6789",
  },
}
`;

exports[`tests for crm-comment.asl.json > mock config scenarios > scenario RetryOnServiceExceptionTest > CustomerAddedToFollowup 1`] = `
{
  "Entries": [
    {
      "Detail": {
        "EmailAddress": "jdoe@example.com",
        "Message": "Customer Added for follow up",
      },
      "DetailType": "CustomerAdded",
      "EventBusName": "default",
      "Source": "LocalTestingSource",
    },
  ],
}
`;

exports[`tests for crm-comment.asl.json > mock config scenarios > scenario RetryOnServiceExceptionTest > DetectSentiment 1`] = `
{
  "LanguageCode": "en",
  "Text": "I am glad to sign-up for this service. Looking forward to different options.",
}
`;

exports[`tests for crm-comment.asl.json > mock config scenarios > scenario ValidationExceptionCatchTest > Check Address 1`] = `
{
  "FunctionName": "\${CheckAddressFunctionArn}",
  "Payload": {
    "city": "Columbus",
    "state": "OH",
    "street": "123 Main St",
    "zip": "43219",
  },
}
`;

exports[`tests for crm-comment.asl.json > mock config scenarios > scenario ValidationExceptionCatchTest > Check Identity 1`] = `
{
  "FunctionName": "\${CheckIdentityFunctionArn}",
  "Payload": {
    "email": "jdoe@example.com",
    "ssn": "123-45-6789",
  },
}
`;

exports[`tests for crm-comment.asl.json > mock config scenarios > scenario ValidationExceptionCatchTest > ValidationException 1`] = `
{
  "Entries": [
    {
      "Detail": {
        "Message": "Validation Exception",
      },
      "DetailType": "ValidationException",
      "EventBusName": "default",
      "Source": "LocalTestingSource",
    },
  ],
}
`;
